File: ui/src/commands/cheat.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import { clamp } from 'es-toolkit';
import { invoke } from '@tauri-apps/api/core';

export function cheatSetFood(food: number) {
  food = Math.trunc(Math.max(0, food));
  return invoke<null>('cheat_set_food', { food });
}

export function cheatSetIron(iron: number) {
  iron = Math.trunc(Math.max(0, iron));
  return invoke<null>('cheat_set_iron', { iron });
}

export function cheatSetMaxResources() {
  return invoke<null>('cheat_set_max_resources');
}

export function cheatSetResources(resources: Resources) {
  return invoke<null>('cheat_set_resources', { resources });
}

export function cheatSetStability(coord: Coord, stability: number) {
  stability = clamp(stability, 0.0, 1.0);
  return invoke<null>('cheat_set_stability', { coord, stability });
}

export function cheatSetStone(stone: number) {
  stone = Math.trunc(Math.max(0, stone));
  return invoke<null>('cheat_set_stone', { stone });
}

export function cheatSetWood(wood: number) {
  wood = Math.trunc(Math.max(0, wood));
  return invoke<null>('cheat_set_wood', { wood });
}

File: crates/nil/src/command/cheat.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

use crate::error::Result;
use crate::manager::ManagerExt;
use nil_core::resource::{Food, Iron, Resources, Stone, Wood};
use nil_core::village::{Coord, Stability};
use tauri::AppHandle;

#[tauri::command]
pub async fn cheat_set_food(app: AppHandle, food: Food) -> Result<()> {
  app
    .client(async |cl| cl.cheat_set_food(food).await)
    .await?
    .map_err(Into::into)
}

#[tauri::command]
pub async fn cheat_set_iron(app: AppHandle, iron: Iron) -> Result<()> {
  app
    .client(async |cl| cl.cheat_set_iron(iron).await)
    .await?
    .map_err(Into::into)
}

#[tauri::command]
pub async fn cheat_set_max_resources(app: AppHandle) -> Result<()> {
  app
    .client(async |cl| cl.cheat_set_max_resources().await)
    .await?
    .map_err(Into::into)
}

#[tauri::command]
pub async fn cheat_set_resources(app: AppHandle, resources: Resources) -> Result<()> {
  app
    .client(async |cl| cl.cheat_set_resources(resources).await)
    .await?
    .map_err(Into::into)
}

#[tauri::command]
pub async fn cheat_set_stability(app: AppHandle, coord: Coord, stability: Stability) -> Result<()> {
  app
    .client(async |cl| {
      cl.cheat_set_stability(coord, stability)
        .await
    })
    .await?
    .map_err(Into::into)
}

#[tauri::command]
pub async fn cheat_set_stone(app: AppHandle, stone: Stone) -> Result<()> {
  app
    .client(async |cl| cl.cheat_set_stone(stone).await)
    .await?
    .map_err(Into::into)
}

#[tauri::command]
pub async fn cheat_set_wood(app: AppHandle, wood: Wood) -> Result<()> {
  app
    .client(async |cl| cl.cheat_set_wood(wood).await)
    .await?
    .map_err(Into::into)
}
